src/node/full_node.hpp

```diff
@@ -94,9 +94,8 @@ class FullNode : public std::enable_shared_from_this<FullNode> {
   void init();
   void close();
 
-  template <typename T, typename... ConstructorParams>
-  auto &emplace(std::shared_ptr<T> &ptr, ConstructorParams &&... ctor_params) {
-    ptr = std::make_shared<T>(std::forward<ConstructorParams>(ctor_params)...);
+  template <typename T>
+  auto const &register_s_ptr(std::shared_ptr<T> const &ptr) {
```

Comment:
I was looking at this before. Can you please explain me what is this even good for ?

Reply:
Long ago we had been having a devastating problem in the tests: nodes would randomly not fully terminate when a test finishes, and the next test would then fail on acquiring some unreleased OS resource like network port or rocksdb db lock.
I fixed it by refactoring the node lifecycle, and put this safeguard to make sure we don't have the problem again.
The idea of it is to make sure that all the shared pointers inside full node are released when the node is destroyed.

Reply:
Ok thanks for explanation. I get the reasoning although I do think that this mechanism lost meaning once you fixed problems with shared pointers... But that will be separate task to refactor this probably...